tezos/context/src/kv_store/persistent.rs

```diff
@@ -0,0 +1,559 @@
+// Copyright (c) SimpleStaking, Viable Systems and Tezedge Contributors
+// SPDX-License-Identifier: MIT
+
+use std::{
+    borrow::Cow, collections::hash_map::DefaultHasher, convert::TryInto, hash::Hasher, io::Write,
+};
+
+#[cfg(test)]
+use std::sync::Arc;
+
+use crypto::hash::ContextHash;
+use tezos_timing::{RepositoryMemoryUsage, SerializeStats};
+
+use crate::{
+    gc::NotGarbageCollected,
+    hash::OBJECT_HASH_LEN,
+    initializer::IndexInitializationError,
+    persistent::{
+        file::{get_persistent_base_path, File, FileType},
+        get_commit_hash,
+        lock::Lock,
+        DBError, Flushable, KeyValueStoreBackend, Persistable,
+    },
+    serialize::{
+        deserialize_hash_id,
+        persistent::{self, read_object_length, AbsoluteOffset},
+        DeserializationError, ObjectHeader,
+    },
+    working_tree::{
+        shape::{DirectoryShapeId, DirectoryShapes, ShapeStrings},
+        storage::{DirEntryId, Storage},
+        string_interner::{StringId, StringInterner},
+        working_tree::{PostCommitData, WorkingTree},
+        Object, ObjectReference,
+    },
+    Map, ObjectHash,
+};
+
+use super::{hashes::HashesContainer, HashId, VacantObjectHash};
+
+const FIRST_READ_OBJECT_LENGTH: usize = 4096;
+
+pub struct Persistent {
+    /// Concatenation of all objects
+    ///
+    /// See `serialize::persistent`
+    data_file: File,
+    /// Store shapes.
+    ///
+    /// File format:
+    /// Concatenation of `StringId` (u32):
+    /// [StringId(0), StringId(9), StringId(3), StringId(33), ..]
+    ///
+    shape_file: File,
+    /// Store shapes indexes.
+    ///
+    /// File format:
+    /// Concatenation of `ShapeSliceId` (u64):
+    /// [ShapeSliceId { start: 0, end: 2 }, ShapeSliceId { start: 2, end: 10}, ..]
+    ///
+    /// In this example, and with the documentation of `shape_file`, the first
+    /// shape (start: 0, end: 2) would be composed of the StringId `0`, `9` and `3`
+    ///
+    shape_index_file: File,
+    /// Store commit indexes.
+    ///
+    /// File format:
+    /// [[HashId (u32), offset (u64), context hash (32 bytes)], [..], ..]
+    ///
+    /// For each commit, we store its `HashId`, its offset in `data_file` and its
+    /// context hash
+    ///
+    commit_index_file: File,
+    /// Store small strings.
+    ///
+    /// File format:
+    /// [[length (u8), string (nbytes)], [..], ..]
+    ///
+    /// The strings 'z' and 'abc' would be written as:
+    /// [1, 'z', 3, 'a', 'b', 'c']
+    ///
+    strings_file: File,
+    /// Store bigs strings
+    ///
+    /// File format:
+    /// [[length (u32), string (nbytes)], [..], ..]
+    ///
+    big_strings_file: File,
+
+    hashes: Hashes,
+    shapes: DirectoryShapes,
+    string_interner: StringInterner,
+
+    pub context_hashes: Map<u64, ObjectReference>,
+
+    lock_file: Lock,
```

Comment:
@sebastiencs clippy says this is never read, lets use `_lock_file` instead (or add a clippy annotation, not sure what is the proper way to be explicit about it)

Reply:
@tizoc Done, I've added a clippy annotation